### PR TITLE
tests: environement setup and other refactorings

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

This update the e2e tests suite to use `TestMain` in order to set up a
environment object for the whole test. This object is meant to store
anything that could be useful for the test to gather information from
it and act based on it (skipping, getting the right image, …)

In addition, this extract the clients into their own package, as well
as the log helpers.

It also refactor a bit how multi-arch is handled, by porting most of
the feature into the environment.

There is a more improvements to do based on this, but this is a first step.

/kind misc
/area testing

/cc @barthy1 @ImJasonH @mattmoor 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by barthy1** [comment]
@vdemeester just ping me, when it's ready to run on s390x env :)

**Comment by vdemeester** [comment]
Note on possible follow-up:
- Run more tests in the same namespace. We create a namespace for almost each and every test (even with table-driven ones). It costs a little bit of testing time.
- Add some "requirement" or "skip" *framework* (not necessarly an external one)
- Expose this to other projects (triggers, cli, …)

**Comment by vdemeester** [comment]
```
no child with platform linux/linux
```

Oups :sweat_smile:

**Comment by vdemeester** [comment]
It is ready for review @barthy1 
/cc @afrittoli

File: test/helm_task_test.go [hunk 1]
```diff
@@ -42,13 +45,17 @@ var (
 // TestHelmDeployPipelineRun is an integration test that will verify a pipeline build an image
 // and then using helm to deploy it
 func TestHelmDeployPipelineRun(t *testing.T) {
+	if testEnv.Platform == "linux/s390x" {
+		t.Skipf("skip for s390x architecture")
+	}
```
Comment (barthy1) [inline test/helm_task_test.go:50]: May be it's better to have 1 entry point for different e2e tests to skip, which can be extended for other architectures? For instance in `setupWithCleanup`?

File: test/internal/environment/environment.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		return nil, err
+	}
+	v, err := c.KubeClient.Kube.Discovery().ServerVersion()
+	if err != nil {
+		return nil, err
+	}
+	return &Execution{
+		Clients:     c,
+		Info:        v,
+		Platform:    v.Platform,
```
Comment (barthy1) [inline test/internal/environment/environment.go:61]: I really like this option to get `platform` from actual target k8s cluster, so no need any more to use `TEST_RUNTIME_ARCH`, in case when tests are initiated on machine of one arch, but will be executed on k8s cluster with another arch 👍 
Can you also remove mention about this env variable from [doc](https://github.com/tektoncd/pipeline/tree/master/test#running) ?

File: test/examples_test.go [hunk 3]
```diff
@@ -53,47 +49,23 @@ func getCreatedTektonCRD(input []byte, kind string) (string, error) {
...
-		output = existingImage.ReplaceAll(output, archSpecificImage)
-	}
-	return output, nil
-}
-
 // koCreate wraps the ko binary and invokes `ko create` for input within
 // namespace
 func koCreate(input []byte, namespace string) ([]byte, error) {
-	cmd := exec.Command("ko", "create", "--platform", "linux/"+getTestArch(), "-n", namespace, "-f", "-")
+	cmd := exec.Command("ko", "create", "--platform", testEnv.Platform, "-n", namespace, "-f", "-")
```
Comment (barthy1) [inline test/examples_test.go:68]: After test case "initiate tests on amd64, execute on s390x", learnt that `--kubeconfig` should be added here. However it probably can be separate PR, as this functionality is not available in the initial code.

**Comment by barthy1** [comment]
the e2e and examples tests passed for s390x. Provided just a few suggestions from my side.

File: test/helm_task_test.go (hunk 1 above)
Comment (vdemeester) [reply test/helm_task_test.go:50]: So I want to make the `skip` very explicit *per* tests. Aka, if a test needs to be skip for whatever reason, it needs to be known when reading the test. In `docker/docker` we used the concept of `requirements` (`requirements.IsNetworkAvailable`, …)

File: test/internal/environment/environment.go (hunk 2 above)
Comment (vdemeester) [reply test/internal/environment/environment.go:61]: we can yes :upside_down_face:

File: test/examples_test.go (hunk 3 above)
Comment (vdemeester) [reply test/examples_test.go:68]: @barthy1 I am not sure I understand :thinking:

Comment (barthy1) [reply test/examples_test.go:68]: If `go test` is running with `--kubeconfig`, `ko create` won't have knowledge about the special kubeconfig and will be targeted to default cluster

Comment (vdemeester) [reply test/examples_test.go:68]: @barthy1 ahhh very good point ! Yeah this can be done in a follow-up or now, I'll see if I've got some time to fix

File: test/helm_task_test.go (hunk 1 above)
Comment (barthy1) [reply test/helm_task_test.go:50]: ok, I have to say I like to keep related things in one place, just not to spam the code with `if` statements for different cases (and many archs in the future), however I'd like to believe that all tests will work for all archs 🤞 and there won't be many exclusions required..

Comment (vdemeester) [reply test/helm_task_test.go:50]: example in `docker/docker`: https://github.com/moby/moby/blob/master/integration/container/create_test.go#L74

**Review by sbwsg — APPROVED** [review]

**Comment by vdemeester** [comment]
/close
I'll rework this at some point.
